File: app/src/main/java/com/example/myappcontacts/presentation/contactsmap/presenter/ContactsMapPresenter.java
package com.example.myappcontacts.presentation.contactsmap.presenter;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.example.myappcontacts.App;
import com.example.myappcontacts.business.contactsmap.IContactsMapInteractor;
import com.example.myappcontacts.data.dao.contactsmap.MapMarkersModel;
import com.example.myappcontacts.di.contactsmap.ContactsMapModule;
import com.example.myappcontacts.presentation.contactsmap.view.IContactsMapView;
import com.google.android.gms.maps.model.Marker;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;

@InjectViewState
public class ContactsMapPresenter extends MvpPresenter<IContactsMapView> implements IContactsMapPresenter {

    private CompositeDisposable mDisposer = new CompositeDisposable();
    private Map<Marker, UUID> mMarkersMap;

    @Inject
    IContactsMapInteractor mIContactsMapInteractor;

    public ContactsMapPresenter() {
        App.get().plusContactsMapModule(new ContactsMapModule()).inject(this);
    }

    @Override
    public void init() {
        loadMarkersList();
    }


    private void loadMarkersList() {
        mDisposer.add(
                mIContactsMapInteractor.loadMapMarkersList()
                        .subscribe(this::onLoadSuccess, this::onLoadError));
    }

    private void onLoadSuccess(List<MapMarkersModel> mapMarkersList) {
        Log.i("MY_TAG3", getClass().getSimpleName() + " onLoadSuccess" + mapMarkersList.size());
        getViewState().showMarkers(mapMarkersList);
    }

    private void onLoadError(Throwable throwable) {
        Log.e("MY_TAG3", getClass().getSimpleName() + " onLoadError " + throwable);
    }

    @Override
    public void setShownMarkersMap(Map<Marker, UUID> markersMap) {
        mMarkersMap = markersMap;
    }

    @Override
    public void onMarkerClicked(Marker marker) {
        getViewState().openContact(mMarkersMap.get(marker));
    }

    @Override
    public void onDestroy() {
        mDisposer.dispose();
        App.get().clearContactsMapComponent();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/myappcontacts/presentation/contactsmap/presenter/ContactsMapPresenter.kt
package com.example.myappcontacts.presentation.contactsmap.presenter

import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.example.myappcontacts.App
import com.example.myappcontacts.business.contactsmap.IContactsMapInteractor
import com.example.myappcontacts.data.dao.contactsmap.db.MapMarkersModel
import com.example.myappcontacts.di.contactsmap.ContactsMapModule
import com.example.myappcontacts.presentation.contactsmap.view.IContactsMapView
import com.google.android.gms.maps.model.Marker
import io.reactivex.disposables.CompositeDisposable
import java.util.*
import javax.inject.Inject

@InjectViewState
class ContactsMapPresenter: MvpPresenter<IContactsMapView>(), IContactsMapPresenter {
    private val disposer = CompositeDisposable()
    private lateinit var markersMap: Map<Marker, UUID>

    companion object {
        private val TAG = "TEST"
    }

    @Inject
    lateinit var contactsMapInteractor: IContactsMapInteractor

    init {
        App.get().plusContactsMapModule(ContactsMapModule()).inject(this)
    }

    override fun init() = loadMarkersList()

    private fun loadMarkersList() {
        disposer.add(contactsMapInteractor.loadMapMarkersList()
                        .subscribe({onLoadSuccess(it)}, {onLoadError(it)}))
    }

    private fun onLoadSuccess(mapMarkersList: List<MapMarkersModel>) {
        Log.i(TAG, "${this::class.simpleName} onLoadSuccess ${mapMarkersList.size})")
        viewState.showMarkers(mapMarkersList)
    }

    private fun onLoadError(throwable: Throwable) {
        Log.e(TAG, """${this::class.simpleName} onLoadError $throwable)""")
    }

    override fun setShownMarkersMap(markersMap: Map<Marker, UUID>) {
        this.markersMap = markersMap
    }

    override fun onMarkerClicked(marker: Marker) {
        viewState.openContact(markersMap.getValue(marker))
    }

    override fun onDestroy() {
        disposer.dispose()
        App.get().clearContactsMapComponent()
        super.onDestroy()
    }

}